File: chapter7end/java/aircraft-positions/src/main/java/com/thehecklers/aircraftpositions/PositionRetriever.java
package com.thehecklers.aircraftpositions;

import lombok.AllArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;

import java.io.IOException;
import java.util.List;
import java.util.function.Consumer;

@AllArgsConstructor
@Configuration
public class PositionRetriever {
    private final AircraftRepository repository;
    private final WebSocketHandler handler;

    @Bean
    Consumer<List<Aircraft>> retrieveAircraftPositions() {
        return acList -> {
            repository.deleteAll();

            repository.saveAll(acList);

            sendPositions();
        };
    }

    private void sendPositions() {
        if (repository.count() > 0) {
            for (WebSocketSession sessionInList : handler.getSessionList()) {
                try {
                    sessionInList.sendMessage(
                            new TextMessage(repository.findAll().toString())
                    );
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: chapter7end/kotlin/aircraft-positions/src/main/kotlin/com/thehecklers/aircraftpositions/PositionRetriever.kt
package com.thehecklers.aircraftpositions

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.socket.TextMessage
import java.io.IOException
import java.util.function.Consumer


@Configuration
class PositionRetriever(
    private val repository: AircraftRepository,
    private val handler: WebSocketHandler
) {
    @Bean
    fun retrieveAircraftPositions() = Consumer<List<Aircraft>> {
        repository.deleteAll()

        repository.saveAll(it)

        sendPositions()
    }

    private fun sendPositions() {
        if (repository.count() > 0) {
            for (sessionInList in handler.getSessionList()) {
                try {
                    sessionInList.sendMessage(
                        TextMessage(repository.findAll().toString())
                    )
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }
}
